### PR TITLE
Fix another `mismatched type __e` error

### PR DESCRIPTION
**Pull request by rkusa**
Hi,

I've found another example (similar to #33), which results in the following less helpful error: `mismatched type __e`.

```rust
#[async]
fn foobar() -> Result<(i32, i32), ()> {
    if false {
        return Ok(3);
    }
    Ok((1, 2))
}
```

Error:

```
error[E0308]: mismatched types
 --> <proc-macro source code>:1:1
  |
1 | __e
  | ^^^ expected integral variable, found tuple
  |
  = note: expected type `std::result::Result<{integer}, _>`
             found type `std::result::Result<(i32, i32), ()>`

error[E0271]: type mismatch resolving `<impl futures::__rt::MyFuture<<[generator@src/main.rs:7:39: 1:5 _] as std::ops::Generator>::Return> as futures::Future>::Item == (i32, i32)`
 --> src/main.rs:7:16
  |
7 | fn foobar() -> Result<(i32, i32), ()> {
  |                ^^^^^^^^^^^^^^^^^^^^^^ expected integral variable, found tuple
  |
  = note: expected type `{integer}`
             found type `(i32, i32)`
  = note: required because of the requirements on the impl of `futures::__rt::MyFuture<std::result::Result<(i32, i32), ()>>` for `impl futures::__rt::MyFuture<<[generator@src/main.rs:7:39: 1:5 _] as std::ops::Generator>::Return>`
  = note: the return type of a function must have a statically known size
```

With this PR, the error is:

```
error[E0308]: mismatched types
 --> src/main.rs:9:19
  |
9 |         return Ok(3);
  |                   ^ expected tuple, found integral variable
  |
  = note: expected type `(i32, i32)`
             found type `{integer}`
```

---

**Edit:** I've just applied this PR to a bigger project and it seems to have a bigger effect on the kind of reported errors than expected. For example errors like

```
error[E0271]: type mismatch resolving `<impl futures::__rt::MyFuture<<[generator@src/models/order.rs:324:57: 1:5 conn:_, order:_ _] as std::ops::Generator>::Return> as futures::Future>::Error == (postgres::Error, postgres_pool::Connection)`
   --> src/models/order.rs:324:6
    |
324 | ) -> Result<(Order, Connection), (DbError, Connection)> {
    |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `postgres_pool::Transaction`, found struct `postgres_pool::Connection`
    |
    = note: expected type `(postgres::Error, postgres_pool::Transaction)`
               found type `(postgres::Error, postgres_pool::Connection)`
    = note: required for the cast to the object type `futures::Future<Item=(models::order::Order, postgres_pool::Connection), Error=(postgres::Error, postgres_pool::Connection)>`
```

are gone and replaced with errors that point directly to the lines inside the function that return the wrong type.

### CODE REVIEW

File: testcrate/ui/unresolved-type.stderr [hunk 1]
```diff
@@ -14,5 +14,5 @@ error[E0412]: cannot find type `A` in this scope
    |
    = help: there is an enum variant `futures::future::Either::A`, try using `futures::future::Either`?
 
-error: aborting due to 3 previous errors
+error: aborting due to 4 previous errors
```
Comment (rkusa) [inline testcrate/ui/unresolved-type.stderr]: TBH, I don't know why this change was necessary, especially since the stderr only contains `2` errors?

**Comment by alexcrichton** [comment]
Oh don't worry errors tend to mystify me as well coming out of the compiler... Regardless looks great to me, thanks!
